samples on non-x86 proposal (#34)
openshift/enhancements: multi-arch/samples.md

those other teams in Red Hat, or by IBM, the samples operator should ensure that it

 - does not install any x86 imagestreams, or templates referencing those imagestreams, on non-x86 systems
 - and by extension, always provides valid ClusterOperator statuses during install and upgrade, as this amounts to a no-op on those platforms

And in case the implication is not clear, OpenShift development will not be developing or maintaining 
any non-x86 images for these existing samples.

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift/docs]

## Summary

As part of achieving what is described in the opening section, the samples operator needs
to achieve the desired results within the context of how multi-arch is being achieved for 
the entire product.

Details around how mulit-arch will be achieved across the entire product are still pending,
at least at the time of this writing for the author.

In particular:

>  - will support across the entire product be staged along homogoneous / heterogeneous lines

Reviewer: We're starting with homogenous.  We'll see about heterogenous down the line.

>  - if "homogeneous only" becomes a valid option for the released product, is use of Golang's APIs for determining architecture sufficient

Reviewer: Yup, that's [what container images use][1].

[1]: https://github.com/opencontainers/image-spec/blame/v1.0.1/config.md#L103-L106

>  - if "heterogenous" is an option, the assumption is the list of valid architectures will be provided in a cluster-wide API object, and that samples operator installs for all architectures listed

Author: though a user could further restrict the arch's installed by updating the Architectures field in the samples operator config object to specify only a subset

Reviewer: This could work.  I'd seen it as "samples operator installs for all possible architectures", which wouldn't need a config.  But maybe that's too fat?  Anyhow, we can punt this until we have homogenous working ;).

>  - if "heterogenous" is an option, some notion outside the scope of the cluster operator API determines if a given sample runs on its associated architecture (for example, is a node selector needed on a deployment config or build leveraging a non-x86 imagestream to make sure the execution occurs on the right architecture?)

Reviewer: Yeah, we'd definitely need node selectors or some such by the time we get heterogenous clusters, because at least some images are going to lack support for at least some arches.  But again, we can punt on heterogenous for now.

## Motivation

Multi-arch has been declared a key initiative for v4.x.

### Goals

> x86 samples are not installed in a cluster with no x86 nodes.  Nor are any future non-x86 samples installed in a cluster with only x86 nodes.

Reviewer: Do you have background motivation for this goal?

Reviewer 2: the motivation is "they won't work"

Reviewer: Does it matter if they won't work?  If we get heterogeneous clusters, I think we'll want to provide samples for all arches that exist in the cluster, so you can, for example, launch a Redis container without caring about the target node arch.  However we do that, I don't see a problem installing samples for arches that don't exist in the cluster too.

Reviewer 2: When we are ready to implement/provide heterogeneous clusters we can revisit (I think i made a comment to that effect later in this proposal), until then, providing x86 samples on a non-x86 cluster that can't include x86 nodes is simply leading users into failure....and right now that failure reason is particularly unclear, as well (for which i've opened a BZ).  Basically if you try to run a ppc image on an x86 node, it will go into a crashloop and the logs/error message are not going to make it apparent why.

Note also that imagestreams do not support a mode of "give me a pod running this imagestreamtag and please pick the right architecture for the node i land on".  That's a manifestlist concept, and imagestreamtags represent specific image manifests, not a manifestlist.  So even if we did provide imagestreamtags that imported various architectures, the user would still have to pick the right imagestreamtag and align it with a nodeselector that steered their pod to the right node architecture.

Again, these are all things that need significant more discussion/design when we are prepared to define true heterogeneous support.

Reviewer: I'd phrase this goal as:

> Install available samples for the current architecture. Do not install samples for other architectures while we only support homogenous clusters, to avoid currently-confusing error messages when an image for one architecture is run on a machine with a different architecture.

That motivates the current plan while making it clear that our concerns about samples from other arches are based on current technical limitations that we might resolve in the future.

Author: switched to that wording ... will push momentarily

### Non-Goals

We do not take on porting existing samples to non-x86 architectures.

## Proposal

The samples operator config object already supports the specification of multiple 
architectures.

It currently requires that only a single architecture is specified, and that 
it is x86_64.

In conjunction, the import of content from https://github.com/openshift/library supports 
downloading multiple architectures, and segregating multiple architectures.  And 
the tagging in https://github.com/openshift/library allows providers to designate which
samples go to which platforms.

So enabling the above to allow PPC and Z content if it is available is item one.

But as previously stated, there is currently no such content, so the samples operator

> should do nothing when it comes up in Managed state, and report Available=true, Progressing=false,

Reviewer: And probably set a reason/message on the Available status like "No samples for `$GOARCH`".

Author: added in next push

Degraged=false.

> We also need to ensure that marking the samples operator as Unmanaged or Removed 

Reviewer: What do we need to ensure about this?  Dangling sentence?

Author: removed in next push .... wasn't anything more needed there

There are some interesting details and caveats that will be discussed in the 
implementation details section.

### User Stories [optional]

#### Story 1

#### Story 2

### Implementation Details/Notes/Constraints [optional]

In a somewhat related note, there have been a lot of side conversations regarding the Jenkins imagestream
in particular, as the openshift/jenkins images are included in the 4.1/4.2 payloads.

Clayton weighed in with Ben Parees and they concluded that unless/until the openshift/jenkins images 
are fully removed from the install payload (i.e. the devtools team, say as part of the Jenkins operator work, 
somehow facilitates the removal of the openshift/jenkins images from the payload), the x86 openshift/jenkins 
images will be utilized as part of building the non-x86 install payload.

### Risks and Mitigations

Unexpected outcomes with respect to the overall multi-arch design.

> Unexpected pushback from IBM on the lack of non-x86 samples.

Reviewer: nit: "lack of non-x86 samples" -> "ownership of non-x86 samples"?  To emphasize the fact that we all agree that we want them, so the pushback would probably be about who has to write them.  And maybe drop "from IBM", because non-OpenShift teams within Red Hat are potential candidates too.

Reviewer 2: I would not categorize the risk as "pushback from IBM".  We want to support these platforms.  Suggest rephrasing as "Unexpected increased priority/requirement to include non-x86 samples"

Author: @bparees 's wording part of next push

## Design Details

### Test Plan

### Graduation Criteria

#### Examples

##### Dev Preview -> Tech Preview

##### Tech Preview -> GA 

##### Removing a deprecated feature

### Upgrade / Downgrade Strategy

No caveats or issues.  Standard, existing approaches apply.

### Version Skew Strategy

No caveats or issues.  Standard, existing approaches apply.

## Implementation History

## Drawbacks

## Alternatives

Remove samples operator as a CVO operator.  Make it a Day 2, Day X operation, where all of the existing samples 
today are installed via OLM Operator Hub.  The selections on the Hub can then explicitly note whether or not they
provide non-x86 options.

## Infrastructure Needed [optional]

New PR e2e jobs running on non-x86, or hybrid topologies, where clear means for the tests to determine the 

> architectures employed such that common tests can be run across multiple architectures.

Reviewer: I hear that something in this space will happen, but am not yet clear on the details.

---
title: Samples-Non-X86
authors:
  - "@gabemontero"
reviewers:
  - "@smarterclayton"
  - "@crawford"
  - "@bparees"
  - "@adambkaplan"
approvers:
  - "@smarterclayton"
  - "@crawford"
  - "@bparees"
creation-date: 2019-09-20
last-updated: 2019-09-20
status: implementable
see-also:
replaces:
superseded-by:
---

# Sample Imagestream and Templates on non-x86 Platforms

As of this writing, all samples are provided and owned by non-OpenShift 
Development teams within Red Hat.  And none of those teams 

>  - provide host images on registry.redhat.io or quay.io that support PPC or Z 

Reviewer 2: i think a some are being built for PPC, e.g.:
https://access.redhat.com/containers/?tab=overview#/registry.access.redhat.com/rhoar-nodejs/nodejs-8

Available CPU Architectures | AMD64PPC64LE

but certainly not Z.

Author: Yeah but they are not included in openshift/library ... I'll clarify

 - declare imagestreams in https://github.com/openshift/library that could reference such images

> OpenShift is now looking at providing support for the two non-x86 architectures from IBM in v4.x.

Reviewer 2: ```suggestion
OpenShift is now looking at providing support for two non-x86 architectures in v4.x.
```

Author: part of next push

> Given that, until a resonable set of samples are provided for the IBM non-x86 architectures by either

Reviewer 2: ```suggestion
Given that, until a resonable set of samples are provided for those architectures, the samples operator should ensure that it
```

Author: part of next push

> those other teams in Red Hat, or by IBM, the samples operator should ensure that it

Reviewer 2: ```suggestion
```

 - does not install any x86 imagestreams, or templates referencing those imagestreams, on non-x86 systems

>  - and by extension, always provides valid ClusterOperator statuses during install and upgrade, as this amounts to a no-op on those platforms

Reviewer 2: ```suggestion
 - and by extension, always provides available/non-degraded ClusterOperator statuses during install and upgrade, as this amounts to a no-op on those platforms
```

any non-x86 images for these existing samples.

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift/docs]

## Summary

As part of achieving what is described in the opening section, the samples operator needs
to achieve the desired results within the context of how multi-arch is being achieved for 
the entire product.

Alex Crawford and Trevor King have indicated that 
 - clusters would remain homogenous at least through 4.5/4.6 at the time of this writing
 - use of golang's runtime.GOARCH is sufficient for determining one's architecture

That said, the samples operator config object currently has a []string array defined "Architectures"
in anticipation of heterogenous clusters.  If that requirement ever gets into plan, the current understanding
after talking to Alex and Trevor are
 - some sort of list of architectures in a cluster would be provided in the infrastructure config

>  - k8s styled selectors of some sort will be provided/utilized to pin pod leveraging non-x86 images to the required nodes; concerns for this pinning thus would not fall on the samples operator

Reviewer 2: though we would have to make a decision about which imagestreams/templates to install (e.g. the x86 ones or the ppc ones, on a cluster that has both x86 + ppc nodes.  We likely can't install both because of naming collisions.  So we'd probably have to pick one and it's up to the admin to reconfigure things later if they want the other...  but we can cross that bridge when we come to it)

Reviewer 3: The alternative is to dive head-first into supporting manifest lists with ImageStreams - that would truly be a v2 change.

## Motivation

Multi-arch has been declared a key initiative for v4.x.

### Goals

x86 samples are not installed in a cluster with no x86 nodes.  Nor are any future non-x86 samples installed in a cluster with only x86 nodes.

### Non-Goals

> We do not take on porting existing samples to non-x86 architectures.

Reviewer 2: ```suggestion
Devex team porting+supplying existing samples for non-x86 architectures.
```

samples go to which platforms.

So enabling the above to allow PPC and Z content if it is available is item one.

But as previously stated, there is currently no such content, so the samples operator
should do nothing when it comes up in Managed state, and report Available=true, Progressing=false,
Degraged=false.

We also need to ensure that marking the samples operator as Unmanaged or Removed 

There are some interesting details and caveats that will be discussed in the 
implementation details section.

### User Stories [optional]

#### Story 1

#### Story 2

### Implementation Details/Notes/Constraints [optional]

In a somewhat related note, there have been a lot of side conversations regarding the Jenkins imagestream
in particular, as the openshift/jenkins images are included in the 4.1/4.2 payloads.

Clayton weighed in with Ben Parees and they concluded that unless/until the openshift/jenkins images 
are fully removed from the install payload (i.e. the devtools team, say as part of the Jenkins operator work, 
somehow facilitates the removal of the openshift/jenkins images from the payload), the x86 openshift/jenkins 

> images will be utilized as part of building the non-x86 install payload.

Reviewer 2: "However as no Jenkins imagestream will be installed on non-x86 clusters, the x86 image will not actually be exposed/consumed on those platforms, it will effectively be inert content."

### Risks and Mitigations

> 

Reviewer 4: Should we call out here or ack that the solution needs to not prevent changes for mixed clusters in the future or how this is expected to evolve?  Maybe overkill since it was already mentioned, just seems like a constraint.

Author: added

Unexpected outcomes with respect to the overall multi-arch design.

Unexpected pushback from IBM on the lack of non-x86 samples.

## Design Details

### Test Plan

### Graduation Criteria

#### Examples

##### Dev Preview -> Tech Preview

##### Tech Preview -> GA 

##### Removing a deprecated feature

### Upgrade / Downgrade Strategy

No caveats or issues.  Standard, existing approaches apply.

### Version Skew Strategy

No caveats or issues.  Standard, existing approaches apply.

## Implementation History

## Drawbacks

## Alternatives

> Remove samples operator as a CVO operator.  Make it a Day 2, Day X operation, where all of the existing samples 

Reviewer 4: might be useful for others to see why the alternative was not preferred to the recommended proposal above.

Author: Containability is the issue here ... and consensus as the long term direction.  I'll try to put in something along those lines.

today are installed via OLM Operator Hub.  The selections on the Hub can then explicitly note whether or not they

> provide non-x86 options.

Reviewer 2: Another alternative is to bootstrap the samples operator as "removed" on non-x86 platforms.  But I don't think that has any meaningful advantages over installing it as managed w/ no imagestreams/templates to install, and i can think of at least a few disadvantages.

### Goals

Install available samples for the current architecture. Do not install samples for other architectures while we 
only support homogenous clusters, to avoid currently confusing error messages when an image for one architecture 
is run on a machine with a different architecture.

### Non-Goals

Devex team porting and supplyling existing samples for non-x86 architectures.

## Proposal

The samples operator config object already supports the specification of multiple 
architectures.

It currently requires that only a single architecture is specified, and that 
it is x86_64.

In conjunction, the import of content from https://github.com/openshift/library supports 
downloading multiple architectures, and segregating multiple architectures.  And 
the tagging in https://github.com/openshift/library allows providers to designate which
samples go to which platforms.

So enabling the above to allow PPC and Z content if it is available is item one.

But as previously stated, there is currently no such content, so the samples operator
should do nothing when it comes up in Managed state, and report Available=true, Progressing=false,

> Degraged=false.

Reviewer 2: ```suggestion
Degraded=false.
```

images will be utilized as part of building the non-x86 install payload.

However as no Jenkins imagestream will be installed on non-x86 clusters, the x86 image will not actually be exposed/consumed on those platforms, it will effectively be inert content.

### Risks and Mitigations

Unexpected outcomes with respect to the overall multi-arch design.

Unexpected increased priority/requirement to include non-x86 samples.

Ensuring that we do no prevent support for heterogeneous cluster in the future.

## Design Details

### Test Plan

### Graduation Criteria

#### Examples

##### Dev Preview -> Tech Preview

##### Tech Preview -> GA 

##### Removing a deprecated feature

### Upgrade / Downgrade Strategy

No caveats or issues.  Standard, existing approaches apply.

### Version Skew Strategy

No caveats or issues.  Standard, existing approaches apply.

## Implementation History

## Drawbacks

## Alternatives

Remove samples operator as a CVO operator.  Make it a Day 2, Day X operation, where all of the existing samples 
today are installed via OLM Operator Hub.  The selections on the Hub can then explicitly note whether or not they
provide non-x86 options.

> This quite possibly is the long term direction for samples operator.  Such a change cannot be contained at this time.

Reviewer 5: imho, this should be the long term direction.

### Goals

Install available samples for the current architecture. Do not install samples for other architectures while we 

> only support homogenous clusters, to avoid currently confusing error messages when an image for one architecture 

Reviewer 2: homogeneous (you have this typo in a bunch of places i think...i know i make it a lot.  homogeneous also, where you have that)

is run on a machine with a different architecture.

### Non-Goals

> Devex team porting and supplyling existing samples for non-x86 architectures.

Reviewer 2: supplying
